### ISSUE TITLE
[Feature] Add tiled output projection and log-probability backward for FSDP2/AutoModel GRPO

### ISSUE THREAD

**Issue by xinyu-dev** [NVIDIA-NeMo/RL#4169]
## Goal

Match the completed PrimeRL Example 5 setup as closely as supported by NeMo RL + NeMo Gym: full-parameter, text-only Nemotron 3.5 Super GRPO using the same Harbor SimpleQA task split and grader.

| Setting | Completed PrimeRL Example 5 | NeMo attempt |
| --- | --- | --- |
| GPU placement | 16 training + 16 inference GPUs | Matched |
| Agent / task | Verifiers Bash + Harbor SimpleQA | Terminus2NemoGym + the same Harbor task split and grader |
| Parameter / optimizer CPU offload | Both off | Matched |
| Activation CPU offload | On, alongside checkpointing | No equivalent exposed in this AutoModel path; checkpointing on |
| Scheduling | Async; maximum policy age 1 | Async; maximum trajectory age 1; implementation differs |
| Batch / group size | 256 / 8 | Matched target; no dynamic replacement |
| Training parallelism | FSDP2 across 16 training GPUs: DP shard 16, DP replicate 1; TP / CP / EP = 1 / 1 / 1 | Matched |
| Precision | BF16 compute; FP32 stored parameters and optimization | FP32 stored parameters; initially FP32 moments, then BF16 retries |
| LR / betas / weight decay / grad norm cap | 1e-5 / 0.9, 0.999 / 0 / 1 | Matched |
| Context / output / maximum turns | 32K / 4K / 8 | Matched limits; actual conversations differ |
| Inference parallelism | 4 separate vLLM engines × TP4 = 16 GPUs; EP1, CUDA graphs, 0.85 memory fraction | Matched settings; different vLLM versions |
| Output-layer memory | Fused, chunked forward and backward | Full FP32 logits and training log-softmax |
| Validation | 16 before/after; temperature 0; interval 10 | Same settings; failed before final validation |
| Agent / verifier timeout | 600 / 120 seconds | Matched |
| Rollout concurrency | Adaptive 8–128, starting at 64 | Fixed maximum 128 |

FSDP2 shards parameters, gradients and optimizer state across the 16 training
ranks; each rank processes different trajectories. TP/CP/EP = 1 therefore still
uses distributed training, without tensor, context or expert parallelism.

## Environment

- NeMo RL: `6453bfd66bda1572654b679fb7a9194f3033b148` for the source paths below.
- AutoModel 0.7.0, submodule `6ae5ca23895d1e06c81cc6103732a20510833fda`; PyTorch 2.11.0+cu130; Transformers 5.15.1.
- Training workload: eight GB200 nodes, four GPUs/node; 16 training GPUs and 16 inference GPUs.
- NeMo Gym: `c3bac96314a59f28b896f597eb9845d175bb0252`; Terminus2NemoGym, maximum 8 turns, summarization disabled.

## Description

Long trajectories on the AutoModel TP1 training path materialize full-vocabulary FP32 logits and log-softmax before gathering the generated-token log-probabilities. The configured `logprob_chunk_size=2048` does not provide a fused, tiled training output layer on this path.

We need selected-token log-probabilities whose output projection, vocabulary normalization and backward pass can run in tiles, without storing a full sequence × vocabulary FP32 table.

## Steps to reproduce the implementation gap

1. Use FSDP2/AutoModel with TP/CP/EP all 1, clipped GRPO, 32K context and the batch/hardware placement above. Keep model/optimizer CPU offload off and activation checkpointing on.
2. Set `policy.logprob_chunk_size: 2048` and inspect the training path:
   - [AutoModel FSDP2 setup](https://github.com/NVIDIA-NeMo/RL/blob/6453bfd66bda1572654b679fb7a9194f3033b148/nemo_rl/models/automodel/setup.py#L475-L485) sets FP32 output dtype.
   - [Forward processing](https://github.com/NVIDIA-NeMo/RL/blob/6453bfd66bda1572654b679fb7a9194f3033b148/nemo_rl/models/automodel/train.py#L332-L345) receives model logits before loss processing.
   - [Training loss preparation](https://github.com/NVIDIA-NeMo/RL/blob/6453bfd66bda1572654b679fb7a9194f3033b148/nemo_rl/models/automodel/train.py#L640-L648) does not pass `chunk_size`.
   - [The TP1 helper](https://github.com/NVIDIA-NeMo/RL/blob/6453bfd66bda1572654b679fb7a9194f3033b148/nemo_rl/distributed/model_utils.py#L1828-L1888) casts logits to FP32, applies full-vocabulary log-softmax, and gathers next-token scores.
3. For runtime investigation, use long Harbor trajectories and inspect allocations during forward/loss/backward after Adam state is resident. The earlier fixed-microbatch Harbor run failed at the FP32-logits allocation during update 2.

The code path is reproducible by source inspection. The failed-run configs and raw trajectories were removed, so the historical OOM is not supplied as a standalone deterministic reproducer. NeMo **does** use `logprob_chunk_size` for [inference logprob postprocessing](https://github.com/NVIDIA-NeMo/RL/blob/6453bfd66bda1572654b679fb7a9194f3033b148/nemo_rl/models/automodel/train.py#L832-L858); this request concerns the training output layer.

## Actual result and memory explanation

The earlier Harbor run completed **1/3 updates**, then requested **13.88 GiB with 6.75 GiB free** while producing FP32 logits. This differs from the later dynamic-batching failure, which occurred earlier in MoE forward.

At the configured maximum length and vocabulary size, one FP32 score table requires:

```text
32,768 tokens × 131,072 vocabulary entries × 4 bytes = 16 GiB
```

PrimeRL 0.9.0 uses **1,024-token × 8,192-vocabulary tiles**, accumulating normalization over the whole vocabulary and recomputing tiles for backward:

```text
1,024 × 8,192 × 4 bytes = 32 MiB per FP32 score tile
```

These are individual buffer sizes, not whole-run peak memory or measured savings. Tiling preserves sequence length and the full vocabulary; it is not truncation or vocabulary pruning. Chunking log-softmax only after producing full logits leaves the initial large allocation in place.

PrimeRL Example 5 completed **3/3 updates plus final validation** with this fused output layer, packing and activation CPU offload. We did not isolate each optimization's contribution; the harnesses and trajectories also differ.

## Expected behavior

- Provide an exposed FSDP2/AutoModel GRPO path that tiles the output projection and vocabulary normalization and recomputes tiles during backward.
- Preserve selected-token log-probabilities, loss masks, sampling-temperature behavior and gradients; validate against the existing path on identical small inputs.
- Avoid a full sequence × vocabulary FP32 allocation throughout this fused path, including backward.
- Document supported loss/sampling options and the meaning of any tile-size settings.

Prior art: [PrimeRL 0.9.0 output layer](https://github.com/PrimeIntellect-ai/prime-rl/blob/v0.9.0/src/prime_rl/trainer/models/layers/lm_head.py#L111-L200). Separate request: [activation CPU offload #4166](https://github.com/NVIDIA-NeMo/RL/issues/4166).

## Run references

- [Retained findings and reproduction guide (NVIDIA access)](https://gitlab-master.nvidia.com/xiyu/benchling-buidathon-super-nemorl/-/blob/xinyu-dev/super35-nemorl/examples/lyris/super35/README.md). Failed-run configs and raw logs were removed during cleanup; the guide preserves their recorded findings.
- [PrimeRL completed-run config](https://github.com/xinyu-dev/benchling-buidathon-super-primerl/blob/222b361ac85c708fbdddd27ba0011fbfc4026697/outputs/nemotron35/grpo-r3-harbor-smoke/configs/resolved/trainer.json) and [results](https://github.com/xinyu-dev/benchling-buidathon-super-primerl/blob/222b361ac85c708fbdddd27ba0011fbfc4026697/outputs/nemotron35/grpo-r3-harbor-smoke/summary.json) (repository access required).
